README.md
# TYPO3 extension ``compatibility6``

This extension adds some features from TYPO3 v6 that were changed in TYPO3 v7,
but makes the upgrade process easier.

|                  | URL                                                     |
|------------------|---------------------------------------------------------|
| **Repository:**  | https://github.com/FriendsOfTYPO3/compatibility6        |
| **TER:**         | https://extensions.typo3.org/extension/compatibility6/  |


ChangeLog
2017-11-23         [RELEASE] Release of compatibility6 7.6.5 (Nicole Cordes)
2017-11-23 df33684 [FEATURE] Add new stage to automatically publish to TER (Nicole Cordes)
2017-11-23 00fb17f Replace tabs with spaces (Nicole Cordes)
2017-11-23 d05aed0 [TASK] Use stages for Travis CI jobs (Nicole Cordes)
2017-10-29 b53ce98 [TASK] Fix coding style according to php-cs-fixer (Nicole Cordes)
2017-10-29 a3f24c5 [FEATURE] Include php-cs-fixer tests in .travis.yml (Nicole Cordes)
2017-10-29 35bef06 [TASK] Add .gitignore file (Nicole Cordes)
2017-10-29 2b3ae5f [FEATURE] Add .travis.yml to run unit tests automatically (Nicole Cordes)
2017-08-12 8aaac61 [BUGFIX] Change storage_pid TCA to work with MySQL strict mode, #18 (Sybille Peters)
2017-08-09 88f0f81 [BUGFIX] Add missing flexParentDatabaseRow to TcaFlexProcess data provider (Nicole Cordes)
2017-06-23 ac025b5 Support integer-type section definitions (Christian Weiske)
2017-02-07 6891044 wizard_link bugfix (CaptainDD)

2016-10-21         Raised version to 7.6.4 (Benni)
2016-09-07 591dbab Fix composer name (Helmut Hummel)
2016-09-07 8027e8c [BUGFIX] Compatibility6 is not sysext anymore (Gerald)
2016-08-30 7417ccf Set vendor name to friendsoftypo3 (Helmut Hummel)

2016-07-26         Merge pull request #3 from pgampe/patch-1
2016-07-26 e7892ba Updated constraints and raise to version 7.6.3 (Benni)
2016-05-20 2444f98 Add missing trim to FlexFormTabsContainer (Philipp Gampe)

2016-02-16         Update ext_emconf.php (Benni)
2016-02-16 9636762 Update ext_emconf.php (Benni)

2016-02-10         [SECURITY] XSS in content element "Form" (Wouter Wolters)
2015-11-12 09c2dbc [TASK] Add autoload information to ext_emconf.php (Nicole Cordes)
2015-11-06 5a789dc [BUGFIX] Fix TYPO3 dependency to allow TER upload (Helmut Hummel)
2015-11-03 2f024fb [TASK] Remove class aliases from non-existent classes (Benjamin Mack)
2015-11-03 754f619 [TASK] Auto-generate LegacyClassesForIde from alias map file (Helmut Hummel)
2015-10-29 c92cbb3 [TASK] compatibility6: Decouple language labels, styles and images (Michael Oehlhof)
2015-11-03 bd3c153 [TASK] Move openid compatiblity config to extension (Helmut Hummel)
2015-10-09 cda497b [!!!][FEATURE] ElementBrowser refactoring, LinkBrowser API - Part 2 (Markus Klein)
2015-10-31 3fbfa35 [TASK] Remove last DocumentTemplate section calls (Wouter Wolters)
2015-10-30 697c80a [TASK] Remove documentTemplate sections at WebFunctionWizardsBaseController (Ruud Silvrants)
2015-10-30 16d8520 [TASK] Remove DocumentTemplate->spacer() and mark as deprecated (Ruud Silvrants)
2015-10-28 1651d79 [TASK] FormEngine: Move SelectTree processing into a data provider (Morton Jonuschat)
2015-10-25 b6581fb [BUGFIX] FormEngine: Inline getDynamicTabMenu (Christian Kuhn)
2015-10-10 7383721 [TASK] Simplify handling of processedTca for FormEngine (Anja Leichsenring)
2015-10-09 6d51f29 [TASK] Handle pageTs for FormEngine in only one result array key, not two (Anja Leichsenring)
2015-10-06 f3d1f2e [TASK] Use ModuleTemplate API for ext:reports (Ruud Silvrants)
2015-10-09 a40b698 [TASK] Make TYPO3 Core PSR-2 standard compliant (TYPO3 Release Team)
2015-10-02 38f7382 [!!!][FEATURE] ElementBrowser refactoring, LinkBrowser API - Part 1 (Markus Klein)
2015-10-06 58a87f7 [TASK] Remove helhum/class-alias-loader in favor of typo3/class-alias-loader (Helmut Hummel)
2015-10-05 e93fc9c [TASK] Call explicit render() on icon objects (Frank Nägler)
2015-09-15 9d7a8c8 [TASK] Cleanup TCA select field configuration (Alexander Stehlik)
2015-09-27 c22b256 [TASK] Extract flex form language handling to compatibility6 (Christian Kuhn)
2015-09-29 6cf4919 [TASK] Update TYPO3 version to TYPO3 7.6-dev (Wouter Wolters)
2015-09-28 90fc7eb [TASK] Replace EXT:t3skin references for icons where possible (Frank Nägler)
2015-09-23 cadc79c [BUGFIX] Conflicting mailform/FORM element in form and compatibility6 (Oliver Hader)
2015-09-21 7070ccb [TASK] Change typo3.org links to https (Josef Glatz)
2015-09-13 dac6999 [TASK] Move content icons to IconFactory using SVG (Stephan Großberndt)
2015-09-15 118ec15 [!!!][TASK] Deprecate IconUtility (Frank Nägler)
2015-06-11 1ec58e3 [CLEANUP] Fix class alias maps and IDE files (Helmut Hummel)
2015-09-15 2976f7f [FEATURE] Add Dispatcher for Backend Routing (Benjamin Mack)
2015-09-14 6d2b421 [TASK] Replace getSpriteIcon in compatibility6 (Michael Oehlhof)
2015-09-11 06ec458 [TASK] Add icon deprecation and unified refresh icon (Benjamin Kott)
2015-08-24 2e042f5 [TASK] Simplify icon handling in FormEngine (Frank Nägler)
2015-09-11 8e0ae9b [TASK] Drop TCA ctrl typeicons (Christian Kuhn)
2015-08-14 f9e39cb [TASK] Ignore composer.json dependencies for TYPO3 Extensions (Helmut Hummel)
2015-07-23 6d00bfe [!!!][TASK] The FormEngine is dead, long live the FormEngine! (Christian Kuhn)
2015-09-07 abe1796 [TASK] Replaced icon with IconFactory: actions-move-* (Frank Nägler)
2015-09-07 b331993 [TASK] Replace icon by the IconFactory actions-system-refresh (Frank Nägler)
2015-09-03 d1b662e [TASK] Replace sprite icon "actions-document-save-close" with IconFactory (Michael Oehlhof)
2015-09-02 c10cbf9 [TASK] Replace sprite icon "actions-document-save" with the new IconFactory (Michael Oehlhof)
2015-08-19 3b60181 [BUGFIX] Add values to submit buttons (Wouter Wolters)
2015-08-10 ff431d3 [TASK] Replace sprite icon "actions-edit-delete" with IconFactory (Markus Guenther)
2015-08-07 a0bdecd [TASK] Add backend routes for wizards (Benjamin Mack)
2015-08-07 f0d768c [!!!][TASK] Move SqlParser into EXT:dbal (Morton Jonuschat)
2015-08-07 2d5e4a7 [TASK] Update to class-alias-loader 2.x (Helmut Hummel)
2015-08-06 5ab3a47 [TASK] BACK_PATH RecordList and friends (Wouter Wolters)
2015-08-06 8ef0fbe [TASK] Get rid of "clear.gif" usages (Wouter Wolters)
2015-08-06 ea6bb84 [TASK] Remove typo3/gfx (Benjamin Mack)
2015-08-04 25271e5 [TASK] Update TYPO3 version to TYPO3 7.5-dev (Wouter Wolters)
2015-07-22 40d884d [TASK] Streamline addModulePath and addModule registrations (Wouter Wolters)

Configuration/TCA/Overrides/tt_content.php
<?php
defined('TYPO3_MODE') or die();

//Extra fields for the tt_content table
$extraContentColumns = [
    'altText' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_altText',
        'config' => [
            'type' => 'text',
            'cols' => '30',
            'rows' => '3'
        ]
    ],
    'imagecaption' => [
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang.xlf:LGL.caption',
        'config' => [
            'type' => 'text',
            'cols' => '30',
            'rows' => '3',
            'softref' => 'typolink_tag,images,email[subst],url'
        ]
    ],
    'imagecaption_position' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:imagecaption_position',
        'config' => [
            'type' => 'select',
            'renderType' => 'selectSingle',
            'items' => [
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang.xlf:LGL.default_value',
                    ''
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:imagecaption_position.I.1',
                    'center'
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:imagecaption_position.I.2',
                    'right'
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:imagecaption_position.I.3',
                    'left'
                ]
            ],
            'default' => ''
        ]
    ],
    'image_link' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_link',
        'config' => [
            'type' => 'text',
            'cols' => '30',
            'rows' => '3',
            'wizards' => [
                'link' => [
                    'type' => 'popup',
                    'title' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_link_formlabel',
                    'icon' => 'EXT:compatibility6/Resources/Public/Images/wizard_link.gif',
                    'module' => [
                        'name' => 'wizard_link',
                    ],
                    'JSopenParams' => 'width=800,height=600,status=0,menubar=0,scrollbars=1'
                ]
            ],
            'softref' => 'typolink[linkList]'
        ]
    ],
    'image_frames' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames',
        'config' => [
            'type' => 'select',
            'renderType' => 'selectSingle',
            'items' => [
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.0',
                    0
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.1',
                    1
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.2',
                    2
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.3',
                    3
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.4',
                    4
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.5',
                    5
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.6',
                    6
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.7',
                    7
                ],
                [
                    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames.I.8',
                    8
                ]
            ]
        ]
    ],
    'longdescURL' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_longdescURL',
        'config' => [
            'type' => 'text',
            'cols' => '30',
            'rows' => '3',
            'wizards' => [
                'link' => [
                    'type' => 'popup',
                    'title' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_link_formlabel',
                    'icon' => 'EXT:compatibility6/Resources/Public/Images/wizard_link.gif',
                    'module' => [
                        'name' => 'wizard_link',
                    ],
                    'params' => [
                        'blindLinkOptions' => 'folder,file,mail,spec',
                        'blindLinkFields' => 'target,title,class,params'
                    ],
                    'JSopenParams' => 'width=800,height=600,status=0,menubar=0,scrollbars=1'
                ]
            ],
            'softref' => 'typolink[linkList]'
        ]
    ],
    'titleText' => [
        'exclude' => true,
        'label' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_titleText',
        'config' => [
            'type' => 'text',
            'cols' => '30',
            'rows' => '3'
        ]
    ]
];

// Adding fields to the tt_content table definition in TCA
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTCAcolumns('tt_content', $extraContentColumns);

// Add default palettes
$GLOBALS['TCA']['tt_content']['palettes'] = array_replace(
    $GLOBALS['TCA']['tt_content']['palettes'],
    [
        '1' => [
            'showitem' => '
				starttime,
				endtime
			'
        ],
        '2' => [
            'showitem' => '
				imagecols,
				image_noRows,
				imageborder
			'
        ],
        '3' => [
            'showitem' => '
				header_position,
				header_layout,
				header_link,
				date
			'
        ],
        '4' => [
            'showitem' => '
				sys_language_uid,
				l18n_parent,
				colPos,
				spaceBefore,
				spaceAfter,
				section_frame,
				sectionIndex
			'
        ],
        '5' => [
            'showitem' => '
				imagecaption_position
			'
        ],
        '6' => [
            'showitem' => '
				imagewidth,
				image_link
			'
        ],
        '7' => [
            'showitem' => '
				image_link,
				image_zoom
			',
        ],
        '8' => [
            'showitem' => '
				layout
			'
        ],
        '10' => [
            'showitem' => '
				table_bgColor,
				table_border,
				table_cellspacing,
				table_cellpadding
			'
        ],
        '11' => [
            'showitem' => '
				image_compression,
				image_effects,
				image_frames
			',
        ],
        '12' => [
            'showitem' => '
				recursive
			'
        ],
        '13' => [
            'showitem' => '
				imagewidth,
				imageheight
			',
        ],
        '14' => [
            'showitem' => '
				sys_language_uid,
				l18n_parent,
				colPos
			'
        ],
        'image_accessibility' => [
            'showitem' => '
				altText;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:altText_formlabel,
				titleText;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:titleText_formlabel,
				--linebreak--,
				longdescURL;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:longdescURL_formlabel
			',
        ]
    ]
);

// Add palettes from css_styled_content if css_styled_content is NOT loaded but needed for CE's "search" and "mailform"
if (!\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::isLoaded('css_styled_content')) {
    $GLOBALS['TCA']['tt_content']['palettes'] = array_replace(
        $GLOBALS['TCA']['tt_content']['palettes'],
        [
            'visibility' => [
                'showitem' => '
					hidden;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:hidden_formlabel,
					sectionIndex;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:sectionIndex_formlabel,
					linkToTop;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:linkToTop_formlabel
				',
            ],
            'frames' => [
                'showitem' => '
					layout;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:layout_formlabel,
					spaceBefore;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:spaceBefore_formlabel,
					spaceAfter;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:spaceAfter_formlabel,
					section_frame;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:section_frame_formlabel
				',
            ]
        ]
    );
}

\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addFieldsToPalette('tt_content', 'image_settings', 'image_frames;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:image_frames_formlabel');

/**
 * CType "search"
 */
$GLOBALS['TCA']['tt_content']['ctrl']['typeicon_classes']['search'] = 'mimetypes-x-content-form-search';
$GLOBALS['TCA']['tt_content']['types']['search'] = [
    'showitem' => '--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.general;general,
			--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.header;header,
		--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.appearance,
			--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.frames;frames,
		--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.access,
			--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.visibility;visibility,
			--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.access;access,
		--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.behaviour,
			--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.searchform;searchform,
		--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.extended'
];

$GLOBALS['TCA']['tt_content']['palettes']['searchform'] = [
    'showitem' => 'pages;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:pages.ALT.searchform',
];

// check if there is already a forms tab and add the item after that, otherwise
// add the tab item as well
$additionalCTypeItem = [
    'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:CType.I.9',
    'search',
    'content-special-indexed_search'
];

$existingCTypeItems = $GLOBALS['TCA']['tt_content']['columns']['CType']['config']['items'];
$groupFound = false;
$groupPosition = false;
foreach ($existingCTypeItems as $position => $item) {
    if ($item[0] === 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:CType.div.forms') {
        $groupFound = true;
        $groupPosition = $position;
        break;
    }
}

if ($groupFound && $groupPosition) {
    // add the new CType item below CType
    array_splice($GLOBALS['TCA']['tt_content']['columns']['CType']['config']['items'], $groupPosition+1, 0, [0 => $additionalCTypeItem]);
} else {
    // nothing found, add two items (group + new CType) at the bottom of the list
    \TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTcaSelectItem(
        'tt_content',
        'CType',
        ['LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:CType.div.forms', '--div--']
    );
    \TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTcaSelectItem('tt_content', 'CType', $additionalCTypeItem);
}

/**
 * CType "mailform"
 */
if (!\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::isLoaded('form')) {
    $GLOBALS['TCA']['tt_content']['ctrl']['typeicon_classes']['mailform'] = 'mimetypes-x-content-form';
    $GLOBALS['TCA']['tt_content']['columns']['bodytext']['config']['wizards']['forms'] = [
        'notNewRecords' => 1,
        'enableByTypeConfig' => 1,
        'type' => 'script',
        'title' => 'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:bodytext.W.forms',
        'icon' => 'EXT:compatibility6/Resources/Public/Images/wizard_forms.gif',
        'module' => [
            'name' => 'wizard_forms',
            'urlParameters' => [
                'special' => 'formtype_mail'
            ]
        ],
        'params' => [
            'xmlOutput' => 0
        ]
    ];

    \TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTcaSelectItem(
        'tt_content',
        'CType',
        [
            'LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:CType.I.8',
            'mailform',
            'content-elements-mailform'
        ],
        'search',
        'before'
    );

    // set up the fields
    $GLOBALS['TCA']['tt_content']['types']['mailform'] = [
        'showitem' => '
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.general;general,
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.header;header,
				bodytext;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:bodytext.ALT.mailform_formlabel,
			--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.appearance,
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.frames;frames,
			--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.access,
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.visibility;visibility,
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.access;access,
			--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.behaviour,
				--palette--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:palette.mailform;mailform,
			--div--;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:tabs.extended'
    ];
    $baseDefaultExtrasOfBodytext = '';
    if (!empty($GLOBALS['TCA']['tt_content']['columns']['bodytext']['defaultExtras'])) {
        $baseDefaultExtrasOfBodytext = $GLOBALS['TCA']['tt_content']['columns']['bodytext']['defaultExtras'] . ':';
    }
    if (!is_array($GLOBALS['TCA']['tt_content']['types']['mailform']['columnsOverrides'])) {
        $GLOBALS['TCA']['tt_content']['types']['mailform']['columnsOverrides'] = [];
    }
    if (!is_array($GLOBALS['TCA']['tt_content']['types']['mailform']['columnsOverrides']['bodytext'])) {
        $GLOBALS['TCA']['tt_content']['types']['mailform']['columnsOverrides']['bodytext'] = [];
    }
    $GLOBALS['TCA']['tt_content']['types']['mailform']['columnsOverrides']['bodytext']['defaultExtras'] = $baseDefaultExtrasOfBodytext . 'nowrap:wizards[forms]';

    $GLOBALS['TCA']['tt_content']['palettes']['mailform'] = [
        'showitem' => 'pages;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:pages.ALT.mailform, --linebreak--, subheader;LLL:EXT:compatibility6/Resources/Private/Language/locallang_ttc.xlf:subheader.ALT.mailform_formlabel',
    ];
}
